### ISSUE TITLE
[Feature Request]: support IME under native Wayland

### ISSUE THREAD

**Issue by Austaras** [electron/electron#33662]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

[Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=1183262) doesn't support IME under native Wayland before ver 98, and IME only works if `gtk-version=4` is enabled after. However when I pass that flag to electron, it would crash saying `GTK 2/3 symbols detected. Using GTK 2/3 and GTK 4 in the same process is not supported`

### Proposed Solution

No idea

### Alternatives Considered

No idea too

**Comment by harmathy** [comment]
For reference: IME refers to something like [ibus](https://github.com/ibus/ibus) or [fcitx](http://fcitx-im.org).

**Comment by OctopusET** [comment]
I'm working on implementing text-input-v3 to chromium. This issue could be solved with this patch.
https://chromium-review.googlesource.com/c/chromium/src/+/3750452

**Comment by alariej** [comment]
@OctopusET Just to understand what the stand is on Electron/Wayland. Would this patch solve the issues I've listed in this Phosh/squeekboard thread:

https://gitlab.gnome.org/World/Phosh/squeekboard/-/issues/244#note_1581502

**Comment by OctopusET** [comment]
As far as I know, yes.
Sorry for delays, I didn't have extra time. I think I can restart work soon,
Ref:
https://www.apertis.org/concepts/on-screen-keyboard/

**Comment by Silentdoer** [comment]
hi, this bug has been fix?

**Comment by OctopusET** [comment]
@Silentdoer Not yet, I'm currently working on now.

**Comment by stalkerg** [comment]
@OctopusET do you need help?

**Comment by OctopusET** [comment]
@stalkerg not now, but could you test it later?

**Comment by stalkerg** [comment]
@OctopusET sure, one more question - is your chromium patch dependent on gtk4?

**Comment by OctopusET** [comment]
no it isn't. GTK4 dependent patch is already there.

**Comment by stalkerg** [comment]
It should be mentioned https://github.com/electron/electron/issues/33690 here, as I understand, no?

**Comment by OctopusET** [comment]
@stalkerg chromium source has gtk4 patch but I think it doesn't enable on electron.
And this patch has some own bugs.

Patch url: https://chromium-review.googlesource.com/c/chromium/src/+/3220698

**Comment by OctopusET** [comment]
Anyone want to talk about this patch development, please come here.
https://matrix.to/#/#chromium-text-input-v3:matrix.org

**Comment by AllanChain** [comment]
Just realized that KWin has got text-input-v1 support recently which makes input methods work in Chromium and Electron apps. (https://invent.kde.org/plasma/kwin/-/merge_requests/3403)

**Comment by stalkerg** [comment]
@AllanChain is correct, and it's good, but it's anyway "deprecated" protocol.

**Comment by Silentdoer** [comment]
hello, can I ask how you're processing at the moment?

**Comment by Kimiblock** [comment]
While Kwin supports text-input-v1, the majority of compositors support the v3 variant

**Comment by tecix** [comment]
> I'm using Gnome and iBus. In Flatpak Chromium putting `gtk-version=4` still couldn't get IME working.

Try `GTK_IM_MODULE=xim XMODIFIERS=@im=ibus google-chrome --gtk-version=4`, this works for me.

**Comment by sith-on-mars** [comment]
> > I'm using Gnome and iBus. In Flatpak Chromium putting `gtk-version=4` still couldn't get IME working.
> 
> Try `GTK_IM_MODULE=xim XMODIFIERS=@im=ibus google-chrome --gtk-version=4`, this works for me.

Didn't work for me unfortunately. This might be a Flatpak specific issue.

**Comment by C0rn3j** [comment]
@sith-on-mars you can probably put a chromium-flags.conf somewhere into your Flatpak with your desired flags.

**Comment by forza91** [comment]
i dont know which side can solve this problem, so here is the Gnome repo, pls help to leave a comment or like:
https://gitlab.gnome.org/GNOME/mutter/-/issues/3200

**Comment by stalkerg** [comment]
@forza91 just in case - IME input for VSCode not working even under KDE.

**Comment by forza91** [comment]
> @forza91 just in case - IME input for VSCode not working even under KDE.

No, bro. Fcitx5 works in KDE Plasma 5.27 with virtual keyboard is on. Native wayland vscode.  You can check the previous comment in this thread

**Comment by Albrecht0xff** [comment]
For vscode and such electron apps, to enable wayland IME support must specify the `--enable-wayland-ime` flag, However, this is not perfect. Usually we modify the vscode startup script directly, or the `.desktop` file. These files will be overwritten to the default values when using the package manager to install an updated version.

To enable wayland support, we can directly set the ozone platform hint in the environment variable, which will not be overwritten by the package manager. 

However, currently just to enable the wayland input method, every time the app updates to a newer version, i will have to modify the startup script again. 

Could we just add a feature that can detect the platform hint environment variable and enable wayland ime by default if the platform hint is wayland or simply add another environment variable called `ELECTRON_ENABLE_WAYLAND_IME`?

**Comment by teohhanhui** [comment]
> Usually we modify the vscode startup script directly, or the .desktop file. These files will be overwritten to the default values when using the package manager to install an updated version.

You make a local copy of the desktop file in `~/.local/share/applications`. You should never modify non-config (`/etc`) files installed by the package manager.

**Comment by Albrecht0xff** [comment]
> > Usually we modify the vscode startup script directly, or the .desktop file. These files will be overwritten to the default values when using the package manager to install an updated version.
> 
> You make a local copy of the desktop file in `~/.local/share/applications`. You should never modify non-config (`/etc`) files installed by the package manager.

Thanks, I forgot that Desktop Entry can be overwritten:

> ArchWiki:
> User entries take precedence over system entries.

**Comment by Kimiblock** [comment]
Just a notice: this workaround requires a compositor supporting text-input-v1, which GNOME's mutter and many other compositors lack or have problematic implementations.

**Comment by xuanruiqi** [comment]
Yeah, this is a non-solution, not even a workaround. It's just utilizing a feature that Electron already has. Nevertheless, the maintainers of Visual Studio Code really could do nothing about this (besides trying to patch, or even better, send PRs to, Electron), as the problem clearly lies with Electron, and is a systematic issue without easy workarounds.

**Comment by czhang03** [comment]
It seems like chromium has start to recieve implementation on `text-input-v3` from orko at igalia.
https://chromium-review.googlesource.com/q/owner:orko@igalia.com 

Hopefully, this can be implemented quickly and adopt by the electron at downstream.

**Comment by forza91** [comment]
Good news!  After few years Chromium finally support the text-input-v3.

https://issues.chromium.org/issues/40113488?pli=1

Thanks for the contributors! We can wait for Electron moving to M129


![2024-08-20 12-22-37](https://github.com/user-attachments/assets/28f99f92-9152-4559-b873-c05cb98c1c2b)

**Comment by czhang03** [comment]
For those who want to test this feature, it is built into the latest flatpak of Chrome dev (flatpak id `com.google.ChromeDev`), and from my testing, it seems to work pretty well.

To test this feature, the tester will need to set `Wayland text-input-v3` to `Enabled` and `Preferred Ozone platform` to `Auto`/`Wayland` in `chrome://flags/`. It can also be toggled with command line flag `--ozone-platform-hint=auto --enable-wayland-ime --wayland-text-input-version=3`. 

According to the [release schedule of electron](https://www.electronjs.org/docs/latest/tutorial/electron-timelines), we probably will need to wait until M130 to get this feature, which is likely to become stable around late October. Then, we also need to wait for specific apps to adopt the new electron version...

**Comment by s0racat** [comment]
https://github.com/electron/electron/issues/33690

**Comment by jexjws** [comment]
Is there any other way to use text-input-v3 besides adding the launch parameter "--wayland-text-input-version=3" currently?

**Comment by eclairevoyant** [comment]
This seems to "just work" as of current electron/chromium, so this can likely be closed out.
